### DISCUSSION TITLE
React class components not rendering with <React.StrictMode>

### DISCUSSION THREAD

**Discussion by acataluddi** [mobxjs/mobx#3393] [Q&A] [answered]
Hi guys,

I was wondering whether any of you experienced any issue with class-based React components and MobX when the `<React.StrictMode>` is enabled. I've created a quick [CodeSandbox Example](https://codesandbox.io/s/7suusz) to explain the issue.

There is a very basic MobX observable store and the following function which triggers the `enableService()` *action*:

```typescript
window.setTimeout(() => {
  store.enableService().then();
}, 1500);
```

The strange thing is that while functional components get updated as expected (the first "service is enabled string) the class-based one is not (the green one):

![immagine](https://user-images.githubusercontent.com/8874358/167914189-05a41ecd-f065-498c-bcea-a52e16051339.png)

However, by removing the `<React.StrictMode>` wrapper, everything works as expected:

```tsx
// Doesn't work
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);

// Works
root.render(
    <App />
);
```

The same code was working without problems in older version of React even with the `<React.StrictMode>` enabled.

Is it possible there is any bug or it's just me missing something obvious?

Thanks,
Adriano

**Comment by urugator** [comment, marked as answer]
Possibly relates to 
> Stricter Strict Mode: In the future, React will provide a feature that lets components preserve state between unmounts. To prepare for it, React 18 introduces a new development-only check to Strict Mode. React will automatically unmount and remount every component, whenever a component mounts for the first time, restoring the previous state on the second mount. If this breaks your app, consider removing Strict Mode until you can fix the components to be resilient to remounting with existing state.

Source: https://github.com/facebook/react/blob/main/CHANGELOG.md#react-1

**Reply by urugator** [reply]
Yea, the problem is that we do not account for a situation where the same component instance is being reused (re-mounted):
We create reaction during the first render and dispose on unmount. We have to patch `componentDidMount`:  if reaction is disposed -> recreate reaction -> forceUpdate()
@mweststrate

**Reply by acataluddi** [reply]
Hi @urugator, 

thank you for looking at this. Should I convert this to an issue then?

Thanks,
Adriano

**Reply by urugator** [reply]
Done, thanks for the report.
